x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java

```diff
@@ -597,4 +706,10 @@ private static boolean verifyBcryptHash(SecureString text, char[] hash) {
         SECURE_RANDOM.nextBytes(salt);
         return salt;
     }
+
+    private static char[] hashSha512(SecureString text) {
+        MessageDigest md = MessageDigests.sha512();
+        md.update(CharArrays.toUtf8Bytes(text.getChars()));
+        return Base64.getUrlEncoder().withoutPadding().encodeToString(md.digest()).toCharArray();
```

Comment:
> Base64 encoding gives a char array of length 86, while toHexCharArray give a char array of length 128. A trivial security gain.

Can you clarify your thinking here? The quantity of meaningful data is the same, it's just the encoding that's different - I don't think that actually affects the security properties.

Reply:
> The quantity of meaningful data is the same
>
This is true. The length of result from `toHexCharArray` is longer, but its valid character selections are smaller (16) compared to base64. So in theory they should provide exact the same search space.

The "trivial" security gain is to assume:
* A blind enumeration through all possible chars/bits without considering its valid char range, i.e. instead of trying only hexdecimal for the 1024 bits output of `toHexCharArray`, it would try all combination of 0s and 1s.
* Original password is longer than 128 chars. So it is more effecitive to enumerate the hash instead of the original password.

Reply:
My reasoning for not making a comment on this is that:

- for online attacks, it doesn't make any difference. Attacker can only control the pre - hashed version 
- for offline attacks, and given our code is public what matters is the search space of the input to pbkdf2 and these are equivalent in both cases . 
 
That said, your comment about String makes sense ( protecting hashes of common passwords to be dumped from memory) and there's no argument in favor of keeping the base64

So +1